### PR TITLE
Split ClientCall into its own class

### PR DESCRIPTION
**Pull request by carl-mastrangelo**

### CODE REVIEW

**Comment by nmittler** [comment]
@carl-mastrangelo +1 to breaking out the call impl ... might be easier to review if you created a separate PR to just move the code.  After it's committed, you could then rebase this PR on it so it's easier to see the diff.  WDYT?

File: core/src/main/java/io/grpc/ClientCallImpl.java [hunk 1]
```diff
@@ -0,0 +1,396 @@
...
+
+  private void startInternal(final ClientStreamListener streamListener, final Headers headers) {
+    ClientTransport transport;
+    try {
+      transport = transportProvider.get();
+    } catch (RuntimeException ex) {
+      // TODO(carl-mastrangelo): Eventually remove this catch statement once obtainActiveTransport
+      // no longer throws.
+      closeCallPrematurely(streamListener, Status.fromThrowable(ex));
+      return;
```
Comment (ejona86) [inline core/src/main/java/io/grpc/ClientCallImpl.java]: These returns changed behavior. Previously they would prevent the timeout from getting registered, now the timeout is still registered which causes apparent leaks (the timeout could be an hour, for all we know). You should make sure that the timeout doesn't get created is the vast majority of cases.

Comment (carl-mastrangelo) [inline core/src/main/java/io/grpc/ClientCallImpl.java]: I'd call that a bug more than a behavior change :)

I split the retry logic out of this class.  PTAL  (some things are a little weird, based on the intent to add back in retry logic.)

File: core/src/main/java/io/grpc/ClientCallImpl.java [hunk 2]
```diff
@@ -0,0 +1,349 @@
...
+  private final ScheduledExecutorService scheduler;
+
+  // Mutable state
+  private ClientStream stream;
+  private ClientCall.Listener<RespT> callListener;
+  private String userAgent;
+  private volatile ScheduledFuture<?> deadlineCancellationFuture;
+  private State state = State.INIT;
+
+
```
Comment (nmittler) [inline core/src/main/java/io/grpc/ClientCallImpl.java]: kill extra lines.

File: core/src/main/java/io/grpc/ClientCallImpl.java [hunk 3]
```diff
@@ -0,0 +1,349 @@
...
+  interface ClientTransportProvider {
+    /**
+     * @return a client transport, or null if no more transports can be created.
+     */
+    ClientTransport get();
+  }
+
+  private enum State {INIT, STARTED, HALF_CLOSED, CANCELED}
+
+  synchronized ClientCallImpl<ReqT, RespT> setUserAgent(@Nullable String userAgent) {
```
Comment (nmittler) [inline core/src/main/java/io/grpc/ClientCallImpl.java]: Why did you add the synchronization to all of the methods?

File: core/src/main/java/io/grpc/ClientCallImpl.java [hunk 4]
```diff
@@ -0,0 +1,349 @@
...
+      return false;
+    }
+    if (transport == null) {
+      closeCallPrematurely(
+          streamListener, Status.UNAVAILABLE.withDescription("Channel is shutdown"));
+      return false;
+    }
+
+    try {
+      synchronized (this) {
```
Comment (nmittler) [inline core/src/main/java/io/grpc/ClientCallImpl.java]: Don't need synchronization here as you sychronize in `start`.

**Comment by nmittler** [comment]
@carl-mastrangelo finished a first pass.  Not sure you need the extra synchronization.

File: core/src/main/java/io/grpc/ClientCallImpl.java (hunk 4 above)
Comment (carl-mastrangelo) [inline core/src/main/java/io/grpc/ClientCallImpl.java]: This was necessary in the retrying case, since start was called possibly from another thread.  Considering that this isn't incorrect, and that this already holds the lock, I would prefer to leave it in place, so I don't forget it in the followup PR.

File: core/src/main/java/io/grpc/ClientCallImpl.java (hunk 2 above)
Comment (carl-mastrangelo) [inline core/src/main/java/io/grpc/ClientCallImpl.java]: Done.

File: core/src/main/java/io/grpc/ClientCallImpl.java (hunk 4 above)
Comment (nmittler) [inline core/src/main/java/io/grpc/ClientCallImpl.java]: Ok, that's fine.

File: core/src/main/java/io/grpc/ClientCallImpl.java [hunk 5]
```diff
@@ -0,0 +1,347 @@
...
+
+    // Fill out the User-Agent header.
+    headers.removeAll(HttpUtil.USER_AGENT_KEY);
+    if (userAgent != null) {
+      headers.put(HttpUtil.USER_AGENT_KEY, userAgent);
+    }
+    return null;
+  }
+
+  private boolean startInternal(final ClientStreamListener streamListener, final Headers headers) {
```
Comment (ejona86) [inline core/src/main/java/io/grpc/ClientCallImpl.java]: So startInternal wasn't part of the necessary. :-/ The point of doing the move separately was so that we wouldn't have to manually discover unrelated changes like this and holding a lock during transport.newStream.

I guess this will need a full review.

Comment (nmittler) [inline core/src/main/java/io/grpc/ClientCallImpl.java]: @carl-mastrangelo I tend to agree with @ejona86 here. Could you just revert any unrelated changes so that we can just review and commit this quickly?

File: core/src/main/java/io/grpc/ClientCallImpl.java (hunk 3 above)
Comment (carl-mastrangelo) [inline core/src/main/java/io/grpc/ClientCallImpl.java]: See comment below.

File: core/src/main/java/io/grpc/ClientCallImpl.java (hunk 5 above)
Comment (carl-mastrangelo) [inline core/src/main/java/io/grpc/ClientCallImpl.java]: @nmittler  & @ejona86  Made Changes as requested.  PTAL.

**Comment by nmittler** [comment]
@carl-mastrangelo LGTM ... thanks for splitting this out!

**Comment by ejona86** [comment]
@carl-mastrangelo LGTM. Thanks, for splitting it out.
